### ISSUE TITLE
Stack overflow in F# compiler. on net5.0

### ISSUE THREAD

**Issue by KevinRansom** [dotnet/runtime#47663]
With dotnet 5.0 we have seen a new stack overflow introduced in the F# compiler when running on the coreclr.

Repro:
unzip [StackOverflow.zip](https://github.com/dotnet/runtime/files/5896821/StackOverflow.zip)
to your local zip.

The project assumes that a dotnet sdk with net5.0 is installed in c:\program files\dotnet

build this project and run it 
````
dotnet build -t:Rebuild -c release
dotnet run -c release
````
observe the stack overflow.

If the project file is edited to netcoreapp3.1 and a netcoreapp3.1 runtime is installed, then executing the program will succeed without a stack overflow.

**Comment by AndyAyersMS** [comment]
Thanks, this repros. Partial stack trace:
```
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr, Void (IntPtr, IntPtr, IntPtr*), IntPtr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].exprF(System.__Canon, Expr)
   at FSharp.Compiler.Detuple+GlobalUsageAnalysis.foldUnderLambda[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Microsoft.FSharp.Core.FSharpFunc`2<Results,Microsoft.FSharp.Core.FSharpFunc`2<System.__Canon,Results>>, Results, System.__Canon)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr, Void (IntPtr, IntPtr, IntPtr*), IntPtr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].exprF(System.__Canon, Expr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].exprNoInterceptF(System.__Canon, Expr)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr, Void (IntPtr, IntPtr, IntPtr*), IntPtr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].exprF(System.__Canon, Expr)
   at FSharp.Compiler.Detuple+GlobalUsageAnalysis.foldUnderLambda[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](Microsoft.FSharp.Core.FSharpFunc`2<Results,Microsoft.FSharp.Core.FSharpFunc`2<System.__Canon,Results>>, Results, System.__Canon)
```

6.0 seems to hang or is very slow? It is perhaps trying to serialize out this stack trace somewhere.

**Comment by AndyAyersMS** [comment]
Still trying to sort out what goes wrong... the repeating pattern on the stack is 36 frames long:
```
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].exprF(System.__Canon, Expr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].exprNoInterceptF(System.__Canon, Expr)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr, Void (IntPtr, IntPtr, IntPtr*), IntPtr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].exprF(System.__Canon, Expr)
   at FSharp.Compiler.Detuple+GlobalUsageAnalysis.foldUnderLambda[[SPC]](FRR, Results, System.__Canon)
   ** at FSharp.Compiler.Detuple+GlobalUsageAnalysis+targetIntercept@291.Invoke(FRE, Results, DecisionTreeTarget)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].targetF(System.__Canon, DecisionTreeTarget)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr, Void (IntPtr, IntPtr, IntPtr*), IntPtr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].exprF(System.__Canon, Expr)
   at FSharp.Compiler.Detuple+GlobalUsageAnalysis.foldUnderLambda[[SPC]](FRR, Results, System.__Canon)
   ** at FSharp.Compiler.Detuple+GlobalUsageAnalysis+targetIntercept@291.Invoke(FRE, Results, DecisionTreeTarget)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].targetF(System.__Canon, DecisionTreeTarget)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr, Void (IntPtr, IntPtr, IntPtr*), IntPtr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].exprF(System.__Canon, Expr)
   at FSharp.Compiler.Detuple+GlobalUsageAnalysis.foldUnderLambda[[SPC]](FRR, Results, System.__Canon)
   ** at FSharp.Compiler.Detuple+GlobalUsageAnalysis+targetIntercept@291.Invoke(FRE, Results, DecisionTreeTarget)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].targetF(System.__Canon, DecisionTreeTarget)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr, Void (IntPtr, IntPtr, IntPtr*), IntPtr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].exprF(System.__Canon, Expr)
   at FSharp.Compiler.Detuple+GlobalUsageAnalysis.foldUnderLambda[[SPC]](FRR, Results, System.__Canon)
   ** at FSharp.Compiler.Detuple+GlobalUsageAnalysis+targetIntercept@291.Invoke(FRE, Results, DecisionTreeTarget)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].targetF(System.__Canon, DecisionTreeTarget)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr, Void (IntPtr, IntPtr, IntPtr*), IntPtr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].exprF(System.__Canon, Expr)
   at FSharp.Compiler.Detuple+GlobalUsageAnalysis.foldUnderLambda[[SPC]](FRR, Results, System.__Canon)
   ** at FSharp.Compiler.Detuple+GlobalUsageAnalysis+targetIntercept@291.Invoke(FRE, Results, DecisionTreeTarget)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].targetF(System.__Canon, DecisionTreeTarget)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr, Void (IntPtr, IntPtr, IntPtr*), IntPtr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].exprF(System.__Canon, Expr)
   at FSharp.Compiler.Detuple+GlobalUsageAnalysis.foldUnderLambda[[SPC]](FRR, Results, System.__Canon)
   ** at FSharp.Compiler.Detuple+GlobalUsageAnalysis+targetIntercept@291.Invoke(FRE, Results, DecisionTreeTarget)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].targetF(System.__Canon, DecisionTreeTarget)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr, Void (IntPtr, IntPtr, IntPtr*), IntPtr)
   at FSharp.Compiler.TypedTreeOps+ExprFolders`1[[SPC]].exprF(System.__Canon, Expr)
   at FSharp.Compiler.Detuple+GlobalUsageAnalysis.foldUnderLambda[[SPC]](FRR, Results, System.__Canon)
   at System.Runtime.CompilerServices.RuntimeHelpers.DispatchTailCalls(IntPtr, Void (IntPtr, IntPtr, IntPtr*), IntPtr)

where

SPC = corelib
FRR = Microsoft.FSharp.Core.FSharpFunc`2<Results,Microsoft.FSharp.Core.FSharpFunc`2<System.__Canon,Results>>
FRE = Microsoft.FSharp.Core.FSharpFunc`2<Results,Microsoft.FSharp.Core.FSharpFunc`2<Expr,Results>>
```
My guess based on some of the comments over in #40864 is that the delegate invokes (`**` in above) could be a factor.

**Comment by AndyAyersMS** [comment]
cc @jkotas @jakobbotsch @janvorli

**Comment by KevinRansom** [comment]
Ouch!!!  I had a similar issue trying to figure it out from the f# source.  Everything looked idiomatic and normal.

**Comment by AndyAyersMS** [comment]
Delegate invoke codegen for 5.0.
```asm
; targetIntercept@291:Invoke

; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; Tier-0 compilation
; MinOpts code
; rbp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 this         [V00    ] (  1,  1   )     ref  ->  [rbp+0x10]   this class-hnd
;  V01 arg1         [V01    ] (  1,  1   )     ref  ->  [rbp+0x18]   class-hnd
;  V02 arg2         [V02    ] (  1,  1   )     ref  ->  [rbp+0x20]   class-hnd
;  V03 arg3         [V03    ] (  1,  1   )     ref  ->  [rbp+0x28]   class-hnd
;  V04 loc0         [V04    ] (  1,  1   )     ref  ->  [rbp-0x08]   must-init class-hnd
;  V05 OutArgs      [V05    ] (  1,  1   )  lclBlk (32) [rsp+0x00]   "OutgoingArgSpace"
;  V06 tmp1         [V06    ] (  1,  1   )     ref  ->  [rbp-0x10]   must-init "argument with side effect"
;
; Lcl frame size = 48

G_M31457_IG01:              ;; offset=0000H
       55                   push     rbp
       4883EC30             sub      rsp, 48
       488D6C2430           lea      rbp, [rsp+30H]
       33C0                 xor      rax, rax
       488945F8             mov      qword ptr [rbp-08H], rax
       488945F0             mov      qword ptr [rbp-10H], rax
       48894D10             mov      gword ptr [rbp+10H], rcx
       48895518             mov      gword ptr [rbp+18H], rdx
       4C894520             mov      gword ptr [rbp+20H], r8
       4C894D28             mov      gword ptr [rbp+28H], r9
						;; bbWeight=1    PerfScore 8.00
G_M31457_IG02:              ;; offset=0024H
       4C8B4D28             mov      r9, gword ptr [rbp+28H]
       4D8B4910             mov      r9, gword ptr [r9+16]
       4C894DF8             mov      gword ptr [rbp-08H], r9
       4C8B4DF8             mov      r9, gword ptr [rbp-08H]
       488B5518             mov      rdx, gword ptr [rbp+18H]
       4C8B4520             mov      r8, gword ptr [rbp+20H]
       48B9A05FB88DFE7F0000 mov      rcx, 0x7FFE8DB85FA0
       E8B5E7FFFF           call     GlobalUsageAnalysis:foldUnderLambda(Microsoft.FSharp.Core.FSharpFunc`2[__Canon,__Canon],Results,System.__Canon):Results
       488945F0             mov      gword ptr [rbp-10H], rax
       488B55F0             mov      rdx, gword ptr [rbp-10H]
       48B99897B58DFE7F0000 mov      rcx, 0x7FFE8DB59798
       E84EF296FC           call     Microsoft.FSharp.Core.FSharpOption`1[__Canon][System.__Canon]:Some(System.__Canon):Microsoft.FSharp.Core.FSharpOption`1[__Canon]
       90                   nop      
						;; bbWeight=1    PerfScore 11.75
G_M31457_IG03:              ;; offset=0063H
       488D6500             lea      rsp, [rbp]
       5D                   pop      rbp
       C3                   ret      
						;; bbWeight=1    PerfScore 2.00

; Total bytes of code 105, prolog size 20, PerfScore 32.25, instruction count 26, allocated bytes for code 105 (MethodHash=cd8c851e) for method targetIntercept@291:Invoke
```

**Comment by AndyAyersMS** [comment]
3.1 has identical codegen here, so perhaps I'm off track.

5.0 also overfflows with TC disabled, so it's not some sort of timing thing where Tier0 frames are larger and 5.0 is taking longer to swap over to Tier1.

**Comment by jkotas** [comment]
Note that the repro stackoverflows on .NET Framework 4.8. I suspect that the regression is caused by fixed stack consumption: some frames got smaller from netfx4.8 to netcoreapp3.1 that made it work, and some frames got bigger from netcoreapp3.1 to net5.0 that broke it again.

**Comment by AndyAyersMS** [comment]
All the user methods in the above 36 frame chunk have identical frame sizes in 3.1 and 5.0. So could be other methods above or below this that are the culprit, or the dispatcher frames. Dispatcher frame is 96 bytes.

Iff it's other methods it suggests the 3.1 version should be close to overflowing.  Running 3.1 under windb, !etb at process exit shows only 0x1B000 stack extent. Doesn't seem likely the dispatcher frames could add up to hundreds of KB. So still puzzled as to where the extra size is coming from.

There are also some odd debugger messages I don't recognize:
```
StackOverflow Information: 0 : Reactor: 44,176.134 enqueue Unknown.Compile (), length 0
StackOverflow Information: 0 : Reactor: 44,176.509 --> Unknown.Compile (), remaining 0
StackOverflow Information: 0 : Reactor: 44,284.273 BAD-OP: >10s <-- Unknown.Compile, took 107757.4633 ms
```
So maybe something in the process is trying to sense potential overflow and adapt?

With 5.0 the stack extent at overflow via !teb is 0x17F000 which jives with the stack reserve in the PE header
```
          180000 size of stack reserve
```

**Comment by AndyAyersMS** [comment]
`StackOverflow Information ...`  looks like this first bit is coming from the app name, so not indicative of some sort internal stack size monitoring.

**Comment by AndyAyersMS** [comment]
Hmm, maybe in 3.1 this computation is not on the main thread.

**Comment by AndyAyersMS** [comment]
Not sure the quickest way to figure max stack extent on a random thread in a process that doesn't crash, but it looks like the worker thread in 3.1 has max stack is somewhere between 0x100000 and 0x140000 (measured by setting data breakpoints in windbg). That is compatible with the idea that the extra overhead from the new tail call support, or some slightly larger frames elsewhere, is the culprit.

**Comment by AndyAyersMS** [comment]
Best guess at present is that on 3.1 this test case is close to overflowing. Am planning on binary searching via editbin to find the maximal sized stack trace for 3.1, then compare that to 5.0 to confirm.

If we do confirm the above, then there's no easy fix we can offer. You would need to modify your code to use less stack.

**Comment by AndyAyersMS** [comment]
Using editbin and some binary searching, the 3.1 app's max stack is between 0x110000 and 0x111000 bytes, and 5.0 is between 0x180000 and 0x181000.

At point of overflow, 3.1 has 8983 active frames; 5.0 has 14612 frames.

Comparing the recurring patterns (with frame sizes in hex) we see things like:

```
;; 3.1
80 @291.Invoke
40 targetF
50 tail call helper stub
B0 foldUnderLambda

;; 5.0
80 @291.Invoke
40 targetF
60 IL_STUB_CallTailCallTarget
60 RuntimeHelpers.DispatchTailCalls
80 exprF
80 foldUnderLambda
```
So there is an additional x70 bytes of stack from stubs and 0x50 from exprF's frame remaining on stack.

Overall what's happening is that each new "run" of tail calls on 5.0 has higher overhead than it did in 3.1 and this case has a recurring pattern where it sets up ~2500 of these runs, and that leads to higher stack usage overall, enough to cause overflow.

**Comment by jkotas** [comment]
Do you see any obvious places where to make these frames smaller? I think this looks by design / won't fix to me.

**Comment by AndyAyersMS** [comment]
> Do you see any obvious places where to make these frames smaller? 

I haven't looked yet. Let me see if there's anything that we can do.

Even if so, the overhead will surely still be larger than in 3.1.

**Comment by AndyAyersMS** [comment]
We might be able to shrink the frame for `DispatchTailCalls` by 0x10 bytes, if we're willing to do a manual EH write through thing. There is already an 8 byte void, so if we can free up one local slot everything can slide down. Will give this a try.

Asking @kunalspathak to look at why this doesn't happen automatically in 6.0.

**Comment by jakobbotsch** [comment]
I believe that we can get rid of `PortableTailCallFrame::Prev` which might help a bit. This is leftover from before the dispatcher was a tight loop, if I remember correctly. The previous frame should already exist as a local in `DispatchTailCall`.

**Comment by AndyAyersMS** [comment]
There is 8 bytes of void in the frame, so seemingly any 8 byte reduction could shrink the whole frame by 16 bytes.

This change decreases the frame size to 0x50, but I am still seeing overflows.

```diff
@@ -319,12 +319,14 @@ public static IntPtr AllocateTypeAssociatedMemory(Type type, int size)
             }
             finally
             {
-                tls->Frame = prevFrame;
+                // Computation duplicated to allow enregistration
+                TailCallTls* tls1 = GetTailCallInfo(callersRetAddrSlot, &callersRetAddr);
+                tls1->Frame = prevFrame;

                 // If the arg buffer is reporting inst argument, it is safe to abandon it now
-                if (tls->ArgBuffer != IntPtr.Zero && *(int*)tls->ArgBuffer == 1 /* TAILCALLARGBUFFER_INSTARG_ONLY */)
+                if (tls1->ArgBuffer != IntPtr.Zero && *(int*)tls1->ArgBuffer == 1 /* TAILCALLARGBUFFER_INSTARG_ONLY */)
                 {
-                    *(int*)tls->ArgBuffer = 2 /* TAILCALLARGBUFFER_ABANDONED */;
+                    *(int*)tls1->ArgBuffer = 2 /* TAILCALLARGBUFFER_ABANDONED */;
                 }
             }
         }
```

**Comment by AndyAyersMS** [comment]
Looks like I was confused above, and the frame size for DispatchTailCalls in 5.0 was 0x50, so above is not any smaller.

@jakobbotsch's suggestion works and reduces the size to 0x40. But the app still overflows, this time with a different stack; that stack still has DispatchTailCalls on it but the remainder of the frames are user code.

So I don't see a way for us to fix this in 5.0. We can take the size reduction for 6.0 as general goodness.

> I think this looks by design / won't fix to me.

Yes, looks that way. Any app that intermixes tail calls like this and has thousands of frames on the stack is going to be vulnerable to changes in frame sizes or implementation details across releases.

**Comment by KevinRansom** [comment]
@AndyAyersMS @jakobbotsch , We will have to do a better job on our end.

Thanks for looking at this, closing

Kevin

**Comment by kunalspathak** [comment]
> Asking @kunalspathak to look at why this doesn't happen automatically in 6.0.

`V04` and `V05` EH-live variables do not get enregistered because they are set to be explicitly zero initialized and we disqualify it [from getting enregistered](https://github.com/dotnet/runtime/blob/366cd23a82fc44ea2c613c36bd4bfdbf8c310e56/src/coreclr/jit/lclvars.cpp#L4100). I believe that condition should be `if (varDsc->lvEhWriteThruCandidate && needsExplicitZeroInit)` meaning if we ever see a variable that need zero initialization + it has its own definition, then we should disqualify it. If the variable just needs zero initialization, we should safely enregister it. I tried that fix on `DispatchTailCalls` and it reduces the stack space from 96 bytes to 88 bytes.

Currently, I am checking out possible diffs that would come out by this change and see if it is worth pursuing, but at least the change is something that we wanted to have "Enregister EH var that are single definition".

**Comment by kunalspathak** [comment]
I did some more experiments and looks like we see regression in code size/perf score.

Today, we just enregister EH vars with single def and in future, we have a plan to support more scenarios like EH vars with multiple defs to be enregistered. We do not support multiple defs yet because it adds register pressure and we end up generating worse code (by having some resolution movs for EH vars). With that model, it will not be correct to enregistering a variable that is just a single definition but needs explicit zero initialization because that is technically multi-def variable. It worked out in this scenario, but most of the time, I saw that the zero initialization happens in prolog and later, there is another write to the stack location of the EH variable. That makes code worse because of extra moves and increased register requirements. Perhaps, once we fine tune register allocator, we can revisit enabling it for zero inits.
